File: src/main/java/net/lumalyte/trivia/models/Question.java
package net.lumalyte.trivia.models;

import net.lumalyte.trivia.util.MessageUtil;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Question {
    private final String question;
    private final String correctAnswer;
    private final List<String> incorrectAnswers;
    private final String category;
    private final String difficulty;
    private final String type;
    private final List<String> shuffledAnswers;
    private final int correctAnswerIndex;

    public Question(String question, String correctAnswer, List<String> incorrectAnswers,
                   String category, String difficulty, String type) {
        this.question = question;
        this.correctAnswer = correctAnswer;
        this.incorrectAnswers = incorrectAnswers;
        this.category = category;
        this.difficulty = difficulty;
        this.type = type;

        // Create and shuffle answers
        this.shuffledAnswers = new ArrayList<>(incorrectAnswers);
        this.shuffledAnswers.add(correctAnswer);
        Collections.shuffle(this.shuffledAnswers);
        this.correctAnswerIndex = this.shuffledAnswers.indexOf(correctAnswer);
    }

    public String getQuestion() {
        return question;
    }

    public String getCorrectAnswer() {
        return correctAnswer;
    }

    public List<String> getIncorrectAnswers() {
        return incorrectAnswers;
    }

    public String getCategory() {
        return category;
    }

    public String getDifficulty() {
        return difficulty;
    }

    public String getType() {
        return type;
    }

    public List<String> getShuffledAnswers() {
        return shuffledAnswers;
    }

    public boolean isCorrectAnswer(String answer) {
        // Strip color codes and convert answer to lowercase for case-insensitive comparison
        answer = MessageUtil.stripColor(answer.toLowerCase().trim());

        // Handle true/false questions
        if (type.equalsIgnoreCase("boolean")) {
            // Convert T/F to True/False
            if (answer.equals("t")) answer = "true";
            if (answer.equals("f")) answer = "false";
            return correctAnswer.equalsIgnoreCase(answer);
        }

        // Handle multiple choice questions
        if (answer.length() == 1) {
            int index = answer.charAt(0) - 'a';
            return index == correctAnswerIndex;
        }

        // Direct answer comparison
        return correctAnswer.equalsIgnoreCase(answer);
    }

    public String getFormattedAnswers() {
        StringBuilder sb = new StringBuilder();
        
        // For true/false questions, always show True as A and False as B
        if (type.equalsIgnoreCase("boolean")) {
            sb.append("\n&eA) &fTrue");
            sb.append("\n&eB) &fFalse");
            return sb.toString();
        }

        // For multiple choice, show all shuffled answers
        for (int i = 0; i < shuffledAnswers.size(); i++) {
            sb.append("\n&e").append((char)('A' + i)).append(") &f")
              .append(shuffledAnswers.get(i));
        }
        return sb.toString();
    }

    public String getCorrectAnswerLetter() {
        // For true/false questions, determine if True is correct (A) or False is correct (B)
        if (type.equalsIgnoreCase("boolean")) {
            return correctAnswer.equalsIgnoreCase("true") ? "A" : "B";
        }
        
        // For multiple choice, use the index
        return String.valueOf((char)('A' + correctAnswerIndex));
    }
} 

File: src/main/kotlin/net/badgersmc/trivia/domain/Question.kt
package net.badgersmc.trivia.domain

/** A trivia question with shuffled multiple-choice answers. */
class Question(
    val question: String,
    val correctAnswer: String,
    val incorrectAnswers: List<String>,
    val category: String,
    val difficulty: String,
    val type: String,
) {
    /** All answers in shuffled order. Correct answer is at [correctAnswerIndex]. */
    val shuffledAnswers: List<String>

    /** Index of [correctAnswer] within [shuffledAnswers]. */
    private val correctAnswerIndex: Int

    init {
        val all = mutableListOf<String>()
        all.addAll(incorrectAnswers)
        all.add(correctAnswer)
        all.shuffle()
        shuffledAnswers = all.toList()
        correctAnswerIndex = shuffledAnswers.indexOf(correctAnswer)
    }

    /**
     * Validates a player's answer. Accepts:
     * - Single letter a/b/c/d (case-insensitive) mapping to shuffled answer index
     * - t/f/true/false for boolean questions (maps to A/B)
     * - Direct text match (case-insensitive)
     */
    fun isCorrectAnswer(answer: String): Boolean {
        val normalized = answer.trim().lowercase()

        // True/false questions
        if (type.equals("boolean", ignoreCase = true)) {
            val mapped = when (normalized) {
                "t", "true" -> "true"
                "f", "false" -> "false"
                else -> normalized
            }
            return correctAnswer.equals(mapped, ignoreCase = true)
        }

        // Single-letter answers
        if (normalized.length == 1 && normalized[0] in 'a'..'d') {
            val index = normalized[0] - 'a'
            return index == correctAnswerIndex
        }

        // Direct text comparison
        return correctAnswer.equals(normalized, ignoreCase = true)
    }

    /** Returns MiniMessage-formatted answer options (A) B) C) D) or True/False). */
    val formattedAnswers: String
        get() {
            val sb = StringBuilder()

            if (type.equals("boolean", ignoreCase = true)) {
                sb.append("<newline><yellow>A) <white>True")
                sb.append("<newline><yellow>B) <white>False")
                return sb.toString()
            }

            for (i in shuffledAnswers.indices) {
                sb.append("<newline><yellow>")
                sb.append(('A' + i))
                sb.append(") <white>")
                sb.append(shuffledAnswers[i])
            }
            return sb.toString()
        }

    /** Returns the letter (A-D or A/B) of the correct answer. */
    val correctAnswerLetter: String
        get() {
            if (type.equals("boolean", ignoreCase = true)) {
                return if (correctAnswer.equals("True", ignoreCase = true)) "A" else "B"
            }
            return ('A' + correctAnswerIndex).toString()
        }
}
